cpp/include/rapidsmpf/shuffler/chunk.hpp

```diff
@@ -25,145 +27,335 @@ namespace rapidsmpf::shuffler::detail {
+        // buffer is non-null and the data buffer is ready.
+        return !data_offsets_.empty()
+               && (data_offsets_[n_messages() - 1] == 0 || (data_ && data_->is_ready()));
+    }
+
+    /**
+     * @brief Returns a description of this chunk.
      *
      * @param max_nbytes The maximum size of the chunk data to include.
-     * @param stream CUDA stream used for device memory operations and kernel launches.
+     * @param stream The CUDA stream.
      * @return The description.
      */
     [[nodiscard]] std::string str(
-        std::size_t max_nbytes = 512,
-        rmm::cuda_stream_view stream = cudf::get_default_stream()
+        size_t max_nbytes = 512, rmm::cuda_stream_view stream = cudf::get_default_stream()
     ) const;
 
     /**
-     * @brief Returns true if the chunk is ready for consumption.
-     *
-     * Checks that the gpu_data's CUDA event is ready, if gpu_data contains a valid
-     * buffer. The CUDA event is used to synchronize the chunk's data to ensure
-     * any allocation or copy (e.g., spilling) is complete before the chunk is
-     * consumed. If expected_num_chunks is greater than 0, or gpu_data_size is 0,
-     * the chunk is considered always ready as it should not have any CUDA data
-     * to receive.
+     * @brief Returns a metadata message that represents this chunk.
      *
-     * @return true if the chunk is ready, false otherwise.
+     * @returns The metadata message as a serialized byte vector.
      */
-    [[nodiscard]] bool is_ready() const;
+    [[nodiscard]] std::unique_ptr<std::vector<uint8_t>> serialize() const;
 
   private:
-    /**
-     * @brief Construct a new chunk of a partition.
-     *
-     * @param pid The ID of the partition this chunk is part of.
-     * @param cid The ID of the chunk.
-     * @param expected_num_chunks If not zero, the number of chunks of the partition
-     * expected to get from the sending rank. Ignored when it is zero.
-     * @param gpu_data_size If known, the size of the gpu data buffer (in bytes).
-     * @param metadata The metadata of the packed `cudf::table` that makes up this
-     * chunk.
-     *  @param gpu_data The gpu_data of the packed `cudf::table` that makes up this
-     * chunk.
-     */
+    // constructor
     Chunk(
-        PartID pid,
-        ChunkID cid,
-        std::size_t expected_num_chunks,
-        std::size_t gpu_data_size,
-        std::unique_ptr<std::vector<uint8_t>> metadata,
-        std::unique_ptr<Buffer> gpu_data
+        ChunkID chunk_id,
+        size_t n_messages,
+        std::vector<PartID> part_ids,
+        std::vector<size_t> expected_num_chunks,
+        std::vector<uint32_t> meta_offsets,
+        std::vector<uint64_t> data_offsets,
+        std::unique_ptr<std::vector<uint8_t>> metadata = nullptr,
+        std::unique_ptr<Buffer> data = nullptr
     );
+
+    ChunkID const chunk_id_;  ///< The ID of the chunk.
+    size_t const n_messages_;  ///< The number of messages in the chunk.
```

Comment:
*question*: this information needs to be sent over the wire, but I think it redundantly encodes `meta_offsets_.size()`, is that right? If yes, should we remove it?

Reply:
I agree. We can remove `Chunk::n_messges_` and use the size of either on these vectors 
```c++
       std::vector<PartID> part_ids,
        std::vector<size_t> expected_num_chunks,
        std::vector<uint32_t> meta_offsets,
        std::vector<uint64_t> data_offsets,
```